---
layout: post
title: Programming
cover: code.jpg
categories: posts
permalink: programming
---

## Tutorials
We will have several computational tutorials throughout the course. As the
course progresses, these materials will be posted below. **Note:** *In some
browsers (Firefox primarily), the end of each sentence is clipped off, making
it difficult to read. If this is occurring, please try using another browser
(Chrome, Safari, etc).*

* [Tutorial 0a: Configuring your computer]({{
  site.baseurl}}/code/t0a_setting_up_python.html) \| This tutorial will walk
  you through how to install a Python 3.7 scientific computing environment.

* [Tutorial 0b: Using the Jupyter notebook]({{
  site.baseurl}}/code/t0b_jupyter_notebooks.html) \| This tutorial will teach
  you how to write code and text in Jupyter notebooks for homework submissions.

* [Tutorial 0c: A Primer on Python Syntax]({{
  site.baseurl}}/code/t0c_python_syntax_and_plotting.html) \| This tutorial
  will walk you through the basics of programming in Python.



## Data sets
<!-- Please download the following data sets, unzip them, and place them in your
`bootcamp/data` folder as described in [the setting up Python tutorial]({{
site.baseurl}}/code/t0a_setting_up_python.html). -->

* [Data set 1](http://rpdata.caltech.edu/courses/course_data/ecoli_growth.zip)
  \| Set of phase contrast and fluorescence images of a growing *E. coli.*
  colony.

<!-- * [Data set
  2](http://rpdata.caltech.edu/courses/course_data/lacI_titration.zip) \| A
  large image set of *E. coli* strains with varying copy numbers of the lacI
  repressor molecule. -->

<!-- * [Data set 3](http://rpdata.caltech.edu/courses/course_data/MS2_nuclei.zip) \|
  An image set of 5' and 3' labeled mRNA expressing in a developing
  *Drosophila* embryo. -->

<!-- ## Python utilities
As sometimes syntax can be difficult, we have written a file with a few
functions written in Python that will make some of the in-class exercises less
cumbersome. Please download them below and place them in your root `bootcamp`
folder.

* [pboc_utilities.py]({{ site.baseurl}}/code/pboc_utilities.py) \| Course utilities. -->



## Exercises
As we progress through the course, the code written in class will be posted
here, along with the polished version of the same material.

* [Exercise 1]({{ site.baseurl}}/code/forward_euler.html) \| numerically
   integrating the differential equation for exponential growth. 

* [exercise 2]({{ site.baseurl}}/code/bacterial_growth.html) \| determining
  bacterial growth rate from microscopy data. \[[data
  set](http://rpdata.caltech.edu/courses/course_data/ecoli_growth.zip)\] 
  <!-- \[[in class]({{ site.baseurl}}/code/bacterial_growth_in_class.ipynb)\] -->

* [exercise 3]({{ site.baseurl}}/code/diffusion_via_coin_flips.html) \|
  simulating diffusing particles with coin flips. 

* [exercise 4]({{ site.baseurl}}/code/diffusion_via_coin_flips_continued.html)
  \| continuation of our stochastic treatment of diffusion. -->
  <!-- \[[in class]({{ site.baseurl}}/code/diffusion_via_coin_flips_continued_in_class.html)\]

<!-- * [exercise 5]({{ site.baseurl}}/code/diffusion_1d_and_frap.html) \|
  numerically solving the master equation for diffusion.  -->
  <!-- \[[in class]({{ site.baseurl}}/code/diffusion_1d_and_frap_in_class.html)\] -->
  
<!-- * [exercise 9]({{ site.baseurl}}/code/laci_titration_day3.html) \| completed
  gene expression project, analyzing the effect of laci titration and comparing
  it to the theory.  \[[data
  set](http://rpdata.caltech.edu/courses/course_data/laci_titration.zip)\]
  \[[in class]({{ site.baseurl}}/code/lacI_titration_day3_in_class.html)\] -->

<!-- * [Exercise 6]({{ site.baseurl}}/code/gene_expression_master_equation.html) \|
  Solving the master equation for constitutive gene expression.  -->
  <!-- \[[In class]({{ site.baseurl}}/code/gene_expression_master_equation_in_class.html)\] -->

<!-- * [Exercise 7]({{ site.baseurl}}/code/taylor_series.html) \| Plotting Taylor
  series approximations for increasing number of terms. -->
  <!-- \[[In class]({{ site.baseurl}}/code/taylor_series_in_class.html)\] -->

<!-- * [Exercise 8] \| Simulating the Buri experiment as an illustration of genetic
  drift.  -->
  <!-- \[[In class]({{ site.baseurl}}/code/genetic_drift_in_class.html)\] -->

<!-- * [Exercise 9]({{ site.baseurl}}/code/lacI_titration_day3.html) \| Completed
  gene expression project, analyzing the effect of LacI titration and comparing
  it to the theory.  \[[Data
  set](http://rpdata.caltech.edu/courses/course_data/lacI_titration.zip)\] -->
  <!-- \[[In class]({{ site.baseurl}}/code/lacI_titration_day3_in_class.html)\] -->

<!-- * [Exercise 10]({{ site.baseurl}}/code/central_limit_theorem.html) \|
  Illustrating The Central Limit Theorem computationally.  -->
  <!-- \[[In class]({{ site.baseurl}}/code/central_limit_theorem_in_class.html)\] -->

<!-- * [Exercise 11]({{ site.baseurl}}/code/inferring_p_heads.html) \| Using
  Bayesian inference to predict the probability of getting heads on an unfair
  coin. -->

<!-- * [Exercise 12] \| Plotting the probability of the states in the MWC model for
  increasing ligand concentration.  -->
  <!-- \[[In class]({{ site.baseurl}}/code/plotting_allosteric_states_in_class.html)\] -->

<!-- * [Exercise 13]({{ site.baseurl}}/code/fly_elongation_rate.html) \| Analysis of
  the MS2 constructs to find rate of transcription in fruit flies.  -->
  <!-- \[[In class]({{ site.baseurl}}/code/fly_elongation_rate_in_class.html)\] -->


## External resources
Below is a list of useful online resources for learning the Python programming
language and principles of programming in general.

* [BE/Bi/NB 203 : Programming in the Biological
  Sciences](http://justinbois.github.io/bootcamp/2020/)

* [Be/Bi 103 : Data Analysis in the Biological
  Sciences](http://www.bebi103.caltech.edu)

* [Codecademy course on Python](https://www.codecademy.com/learn/python)

* [DataCamp course on Python for data
  science](https://www.datacamp.com/courses/intro-to-python-for-data-science)
