---
layout: post
title: Programming
cover: code.jpg
categories: posts
permalink: programming
---

## Tutorials
We will have several computational tutorials throughout the term. As the course
progresses, these materials will be posted below. **Note:** *In some browsers (Firefox primarily), the end of each sentence is clipped off, making it difficult to read. If this is occurring, please try using another browser (Chrome, Safari, etc).*

* [Tutorial 0a: Configuring your computer]({{ site.baseurl }}/code/t0a_setting_up_python.html) \| This tutorial will walk you through how to install a Python 3.6 scientific computing environment.

* [Tutorial 0b: Using the Jupyter notebook]({{ site.baseurl }}/code/t0b_jupyter_notebooks.html) \| This tutorial will teach you how to write code and text in Jupyter notebooks for homework submissions.

* [Tutorial 0c: A Primer on Python Syntax]({{ site.baseurl }}/code/t0c_python_syntax_and_plotting.html) \| This tutorial will walk you through the basics of programming in Python.

## Data sets
Please download the following data sets, unzip them, and place them in your `pboc/data` folder as described in [the setting up Python tutorial]({{ site.baseurl }}/code/t0a_setting_up_python.html).

* [Data set 1](http://rpdata.caltech.edu/courses/course_data/ecoli_growth.zip) \| Set of phase contrast and fluorescence images of a growing *E. coli.* colony.

* [Data set 2](http://rpdata.caltech.edu/courses/course_data/lacI_titration.zip) \| A large image set of *E. coli* strains with varying copy numbers of the lacI repressor molecule.

* [Data set 3](http://rpdata.caltech.edu/courses/course_data/ms2_elongation.zip) \| An image set 5' and 3' labeled mRNA expressing in a developing *Drosophila* embryo.

## Python utilities
As sometimes syntax can be difficult, we have written a file with a few functions written in Python that will make some of the in-class exercises less cumbersome. Please download them below and place them in your root `pboc` folder.

* [pboc_utilities.py]({{site.baseurl}}/code/pboc_utilities.py) \| Course utilities.


## Exercises
As we progress through the course, the code written in class will be posted here, along with the polished version of the same material.

* [Exercise 1]({{site.baseurl}}/code/bacterial_growth.html) \| Inferring bacterial growth rate from microscopy data. \[[Data set](http://rpdata.caltech.edu/courses/course_data/ecoli_growth.zip)\] \[[In class]({{site.baseurl}}/code/bacterial_growth_in_class.html)\]

* Exercise 2 \| Modeling diffusion with coin flips. \[[In class]({{site.baseurl}}/code/diffusion_via_coin_flips_in_class.html)\]

* Exercise 3 \| Study of the diffusion timescale. \[[In class]({{site.baseurl}}/code/diffusion_timescale_in_class.html)\]

* Exercise 4 \| Numerical solution to 1-d diffusion and application to FRAP. \[[In class]({{site.baseurl}}/code/diffusion_1d_and_FRAP_in_class.html)\]

* Exercise 5 \| Mean dynamics of constitutive gene expression. \[[In class]({{site.baseurl}}/code/constitutive_expression_in_class.html)\]

* Exercise 6 \| Master equation treatment of constitutive gene expression. \[[In class]({{site.baseurl}}/code/gene_expression_master_equation_approach_in_class.html)\]

* Exercise 7 \| Estimating the mRNA elongation time in *Drosophila* embryos. \[[Data set](http://rpdata.caltech.edu/courses/course_data/ms2_elongation.zip)\] \[[In class]({{site.baseurl}}/code/fly_elongation_rate_in_class.html)\]


## External resources
Below is a list of useful online resources for learning the Python programming language and principles of programming in general.

* [BE/Bi/NB 203 : Programming in the Biological Sciences](http://justinbois.github.io/bootcamp/2017/)

* [Be/Bi 103 : Data Analysis in the Biological Sciences](http://www.bebi103.caltech.edu)

* [Codecademy course on Python](https://www.codecademy.com/learn/python)

* [DataCamp course on Python for data science](https://www.datacamp.com/courses/intro-to-python-for-data-science)
